### ISSUE TITLE
Missing awkward.Array.ndim in the numba context

### ISSUE THREAD

**Issue by tamasgal** [scikit-hep/awkward#572]
I am need some functions which has to work on arrays with arbitrary dimensions, similar to the example below which traverses the `awkward.Array` recursively until it e.g. reaches `ndim == 2` and performs some operations. The problem is that `ak.Array.ndim` is not supported in `numba` and I have not found other ways to travel through dimensions 👨‍🚀. I guess it is just not connected to the numba interface, or is there another way to access the dimensionality which is preferable?
I found some traces of `ndim` in the numba context here: https://awkward-array.readthedocs.io/en/latest/_auto/ak._connect._numba.layout.BitMaskedArrayType.html?highlight=ndim#ak-connect-numba-layout-bitmaskedarraytype-ndim

```python
def mask_startend(arr, start, end):
    builder = ak.ArrayBuilder()
    _mask_startend(arr, start, end, builder)
    return builder.snapshot()

@nb.njit  # raises 
def _mask_startend(arr, start, end, builder):
    if arr.ndim == 2:  # recursion stop
        for el in arr:
            if ak.count(el) > 0 and el[0] == start and el[-1] == end:
                builder.boolean(True)
            else:
                builder.boolean(False)
        return
    
    for subarray in arr:
        builder.begin_list()
        _mask_startend(subarray, start, end, builder)
        builder.end_list()
```

Here are some examples without `njit`:

```python
>>> mask_startend(ak.Array([[1,2,3], [1,2]]), 1, 3)
<Array [True, False] type='2 * bool'>
>>> mask_startend(ak.Array([[[1,2,3], [1,2]], [[1, 3]], [[1, 2], [3, 4]]]), 1, 3).tolist()
[[True, False], [True], [False, False]]
```

and here the error raised with `njit`:

```
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Internal error at resolving type of attribute "ndim" of "arr".
array does not have a field with key 'ndim'

(https://github.com/scikit-hep/awkward-1.0/blob/1.0.0rc2/src/awkward/_connect/_numba/layout.py#L324)
During: typing of get attribute at <ipython-input-135-3d8beb6b636f> (8)
Enable logging at debug level for details.

File "<ipython-input-135-3d8beb6b636f>", line 8:
def _mask_startend(arr, start, end, builder):
    if arr.ndim == 2:  # recursion stop
```

**Comment by jpivarski** [comment]
Missing `ndim` in the Numba context is an oversight, and won't be too hard to write.

(I'm going to start using the phrase "in the Numba context" from now on—it's much better than the "in Numbafied functions," "in functions compiled by Numba," or the not-really-correct "in Numba.")

**Comment by tamasgal** [comment]
After a quick debugging session I found the place where it probably should go:

https://github.com/scikit-hep/awkward-1.0/blob/5996899a390099c9637aafc21961d7a041f1ddf7/src/awkward/_connect/_numba/arrayview.py#L647

It's already marked with `???` 😉 

I am not so familiar with the low level Numba architecture but I'll have a look.

I also found the `@nb.extending.overload_attribute` decorator, maybe that's a better approach than hooking into attributes. No idea...

**Comment by jpivarski** [comment]
That `???` is for future additions, so even after adding `ndim`, there will still be a `???` for other attributes we want to enable in Numba.

I'm not asking you to do it—Numba is a very specialized compiler that wasn't originally designed to be developed beyond the core developers. A few years ago, an extension mechanism was added on, though I still had to delve into the Numba internals to find out how to do basic things, and it was this year reorganized to make extensions easier for outsiders, but that's still evolving. I'm considering giving tutorials on Numba extension development, since there aren't a lot of good materials out there, but I haven't done that yet. That would be for people who have already decided that they want to do significant projects in Numba, like [vector](https://github.com/scikit-hep/vector) or [hist](https://github.com/scikit-hep/hist).

But anyway, here's a general sketch of how it goes: to add a new type, attribute, method, or function to Numba, you have to add type inference, possibly a model, and a lowered implementation. The compiler passes over the code two times (that you care about as an extender—other passes are for other purposes), first to derive output types from input types, and the second to generate LLVM instructions from the Python code. This `???` is in the typing pass: what should go here is

```python
if attr == "ndim":
    return numba.intp
```

(Though I can't remember if this typer should be returning the return type, `numba.intp`, or a type signature, `numba.intp()`. That's something I'd have to look up by examining other code. If it's wrong, you get error messages that don't seem to be related because the Python code that picks up the output of the typer assumes it's the right thing.)

After assigning the type, you would then have to write the code that "lowers" Python bytecode into LLVM assembly. Sometimes, you can do that by writing more Python code and asking Numba to lower it for you, like this:

https://github.com/scikit-hep/awkward-1.0/blob/5996899a390099c9637aafc21961d7a041f1ddf7/src/awkward/_connect/_numba/arrayview.py#L1105-L1171

It creates a string of Python code, based on the inputs, and passes that to Numba in place of the more generic Python we start with. Eventually, though, everything has to resolve into LLVM instructions, and for something as simple as `ndim`, I'd just write the LLVM directly. For the granularity of Numba's type system, the number of dimensions of a NumPy array is part of the type, not a value to be encountered at runtime (they didn't have to do it that way—there's a lot of freedom in designing type systems), and I did Awkward Array the same way: the type at compile time knows the dimension of the array. Thus, if you pass another array with a different number of dimensions to the same function, the function must be recompiled. However, it can run faster once compiled.

So the lowering of `ndim` would be something like

```python
context.get_constant(numba.intp, datatype.get_the_number_of_dimensions)
```

The Python code for getting the number of dimensions from the type information might be involved, but it runs at compile-time and generates an LLVM instruction for a constant integer.

I mentioned a "model" as one of the steps. At runtime, we have LLVM structs instead of Python objects and any runtime information has to be derived from these models. However, the number of dimensions can be determined at compile-time, so we can ignore the Awkward Array's model (called an `ArrayView` in this code).

In summary, it's not for the faint of heart, but if you _want_ to learn how to extend Numba, possibly to use it in another project, then I can guide you through this and you'll learn some things you can use in that other project.

**Comment by tamasgal** [comment]
I was already wondering where your golden source of Numba mechanics is located. Now I understand 😅 

Yes, I already realised that Numba passes over the function multiple times and it makes of course sense, but it's really hard to picture where which types or interpretations are to be passed so that the inference works. The error messages do not help much ;) Therefore I tried to reverse engineer how you do it for other things (fields) but most of them are quite complex array mechanics and not so simple to derive what needs to be created for a simple constant.

> I'm not asking you to do it

I know, I just wanted to see if I can somehow help out since you already have a lot of stuff ongoing. However, it seems that it would be yet another largish project for me to dive in and I am not sure if it's a good idea in this time period 😆

**Comment by jpivarski** [comment]
As it turns out, it was simpler than I was expecting:

https://github.com/scikit-hep/awkward-1.0/pull/578/commits/3d76da9dccade6db7e7d93631c253308fc7bb73c

After adding `if attr == "ndim"` to the typing pass, the lowering pass didn't need another decorated function, just an `if attr == "ndim"` of its own. And another thing I recently added required me to put `ndim` properties on all the type objects, so there wasn't much to do, after all.

There was one subtlety: partitioned arrays are handled separately, so the update had to go into both the ArrayView handler and the PartitionedArrayView handler.

**Comment by tamasgal** [comment]
Thanks Jim, I bookmark this as a reference for similar issues in future. Hopefully I can contribute then ;)

**Comment by jpivarski** [comment]
If you're interested in learning how to extend Numba, I'll keep that in mind for a possible future tutorial. I just think that the audience will be small.  `:)`

**Comment by tamasgal** [comment]
OK thanks, I do! 😉 

Just banging my head against the wall since I ran into the next issue: an extremely limited recursion support in Numba, which makes my example above basically impossible to lower. Now trying to wrap my head to unroll it, but this whole high-performance Python-stack is starting to get a bit annoying nowadays 🙈  Glad that you created the `awkward` package which makes many things possible though!

**Comment by jpivarski** [comment]
The biggest problem for recursion is propagating type information. Often, you get into situations in which the return type depends on the argument type, but the argument type depends on the return type. Since Numba is starting from Python, it usually lacks the type annotations that would resolve this circular dependency. It is, however, possible to add type annotations to the `@numba.jit` operator: https://numba.pydata.org/numba-doc/latest/user/jit.html#eager-compilation

Are you using these?

(Saying "type annotations" suggests that they ought to be Python type annotations, but I don't think Numba knows how to convert Python annotations into Numba annotations. The latter have finer granularity: `int64`, `int32`, `int16`, etc. instead of arbitrary-precision `int`, for instance, but one-way Python annotations → Numba annotations ought to be possible.)

**Comment by tamasgal** [comment]
Yes I have starting looking into them, the problem is that I am not dealing with primitive types but the more complex `awkward.Array`s. In my initial example above you see that the dimension is reduced in each recursion. As you described, it's somehow chocking on the type information provided by the `getiter` functionality. I am already trying to boil it down to an MWE so that I can inspect it more easily but it's hard to reproduce.

**Comment by jpivarski** [comment]
By choosing to make `ndim` part of a NumPy array type, Numba made it so that each level of recursion has to be a separate compilation: a function compiled for a 1-dimensional array can't be used for a 2-dimensional array, etc. Such a thing would have to have overloads (recompilations) for each number of array dimensions. Furthermore, the choice of which overload to use would have to be made using type information, not a runtime thing like an `if` statement. If you had

```python
if array.ndim == 1:
    return do_base_case(array)
else:
    return do_recursive_step(array)
```

then both parts of the `if` statement get _compiled_ for all `ndim` cases, even though some of them will never be reached at runtime. Making `if` statements "narrow" the type in the two branches gets complicated fast: you have to recognize that the `if` predicate is fully known at compile-time, and that predicate can be any expression.

You want a compile-time `if`, but of course Python doesn't have anything like that.

What _should_ be possible is to use [@numba.generated_jit](https://numba.pydata.org/numba-doc/dev/user/generated-jit.html) to make a compile-time decision based on type and provide different implementations for each type, such as arrays of different dimension. This _should_ be possible—all of the information is there:

```python
>>> @nb.generated_jit(nopython=True)
... def do_recursive(arg):
...     # arg is a type (instance of nb.types.Array)
...     if arg.ndim == 1:
...         # arg is a value (array at runtime)
...         return lambda arg: arg
...     else:
...         # arg is a value again (different type, though)
...         return lambda arg: do_recursive(arg[0])
... 
>>> array = np.arange(2*3*5).reshape(2, 3, 5)
>>> array
array([[[ 0,  1,  2,  3,  4],
        [ 5,  6,  7,  8,  9],
        [10, 11, 12, 13, 14]],

       [[15, 16, 17, 18, 19],
        [20, 21, 22, 23, 24],
        [25, 26, 27, 28, 29]]])
>>> do_recursive(array)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
...
NotImplementedError: Failed in nopython mode pipeline (step: nopython frontend)
call to CPUDispatcher(<function do_recursive at 0x7f5286bf6ca0>): unsupported recursion
```

That looks like work TBD. Unlike the `if` branches that depend on type, there aren't fundamental issues for this one, it _should_ work. This is where it should be implemented:

https://github.com/numba/numba/blob/1041fa6ee8430471da99b54b3428a673033e7e44/numba/core/typeinfer.py#L1578-L1587

The part that is implemented depends on whether a `callframe` can be found in `context.callstack`... I don't know enough about this.

**Comment by tamasgal** [comment]
Ah I see, thanks, that's very helpful. There are some significant differences on how Numba is dealing with type inference compared to Julia, which I am familiar with. Yes I already saw that the dimensionality is part of the type, which requires separate compilations.
The [@numba.generated_jit](https://numba.pydata.org/numba-doc/dev/user/generated-jit.html) is interesting, I have not seen that before.

Btw. I opened an issue meanwhile since it seems that parts of the type inference are somehow working, at least in that recursive function but it seems that something might be missing from the awkward side. Not sure though. https://github.com/scikit-hep/awkward-1.0/issues/580

Anyways, please see all these as low priority, I am already working on it in parallel and I guess it's quite a special use-case anyways. If I fail, I will just implement some hardcoded routines for `ndim={4,3,2}`, which should cover most of our users needs.
